plutus-core/docs/equalsCanonical.md

```diff
@@ -0,0 +1,299 @@
+
+> It is worth noting that the presence of effects -- such as nontermination, mutable state, control effects -- requires different formulations of the logical relation that defines parametricity. However, those logical relations capture parametricity in that they always formalize uniformity of behavior across different type instantiations. For instance, for a language that supports nontermination, the logical relation for parametricity ensures that two different instantiations have the same termination behavior: either both diverge, or they both terminate with related values. Because of this, the presence of effects usually leads to weaker free theorems -- in pure System F all inhabitants of `∀𝑋. 𝑋 → 𝑋` are equivalent to the identity function, but in System F with non-termination, every inhabitant is either the identity or always errors. Though the free theorems are weaker, parametricity still ensures uniformity of behavior. As our counterexample above illustrates, GSF is non-parametric since it does not ensure uniform behavior. However, since the difference in behavior was between error and termination, it is possible that GSF satisfies a property that could be called łpartial parametricityž (or parametricity modulo errors) that weakens the notion of uniformity of behavior: either one side errors or we get related behaviors.
+
+We have non-termination, `error` and `trace` in Plutus meaning we cannot use some very general notion of parametricity anyway, although admittedly `equalsCanonical` is far more "malicious" than those effects. Even then, there do exist notions of parametricity for systems even with full-scale type case, quoting [Generalizing Parametricity Using Information-flow](https://www.seas.upenn.edu/~sweirich/papers/seckinds/lics2005.ps):
+
+> Currently, module implementors rely on parametric polymorphism to provide integrity and confidentiality guarantees about their abstract datatypes. However, standard parametricity does not hold for languages with run-time type analysis; this paper shows how to generalize parametricity so that it does. The key is to augment the type system with annotations about information-flow. Implementors can then easily see which parts of a program depend on the chosen implementation by tracking the flow of dynamic type information.
+
+They use the same structural equality function as a "canonical example of run-time type analysis".
+
+Overall, a structural equality operator does complicate metatheory, but it doesn't destroy it.
+
+It is however important to look at the whole system, which in our case comprises four languages, not just TPLC and PIR.
+
+## Security of the system as a whole
+
+If `equalsCanonical : all a. a -> a -> bool` is still admissible, do we win anything at all by introducing the `mkCanonical` indirection? Yes: we introduce a hoop that the user needs to jump through to be able to use `equalsCanonical` and that hoop acts as explicit guidance.
```

Comment:
> I don't find this convincing. Plutus Core is the low-level language, it's okay for people to have to know what they're doing. 

We assume that "people" here are compiler developers. In our case these are the same compiler developers who introduced the "pointless" `Normalized` `newtype` wrapper that easily can be and regularly does get bypassed. `canonical` is very much just like `Normalized`, it's too is a thin wrapper that one can create and remove willy-nilly if they wish so, except in the case of `canonical` there's also a trans-language ghost of a departed proof making it harder to shoot oneself in the foot. Do you think `Normalized` wasn't a good idea? Do you think `canonical` is different?

> it's okay for people to have to know what they're doing. 

I mean, one can go pretty far with this argument. Why do we need static types if we expect people to know what they're doing? The answer is "explicit guidance is helpful". And it is helpful with `Normalized` the same way as it's helpful with `canonical`: it helps when you write, code, it helps even more when you refactor it, it helps even morer when you need to change code that you are not familiar with etc etc. I really fail to understand how that is not completely obvious. 

Even without the `a -> data` ghost, I'd still find `canonical` useful. Creating `canonical` values and using them are two different things and I'd rather have this distinction explicitly and annotate calls to `mkCanonical` with a comment on why it's safe there than do that to calls of `equalsCanonical` potentially deep down the road where all relevant context is lost.

Reply:
> Do you think Normalized wasn't a good idea? Do you think canonical is different?

Yes and yes. `canonical` is built into the language. If you (a compiler writer) are worried about getting it wrong in your generated code, you can very easily define such wrapper functions (or we can provide them in the PLC stdlb). I don't see what advantage we get from building it into the builtins themselves.

Reply:
> > Do you think Normalized wasn't a good idea? Do you think canonical is different?
>
> Yes and yes.

I'm really shocked by the first "yes". I guess we won't be able to agree on anything here then. Which is fine, since we aren't planning to implement the proposal anyway.

Reply:
Er, sorry, I just read it backwards. I think it was a good idea!